Use the Errno module rather than matching against the error message
text (which is locale sepecific).

Add "use Errno" where appropriate.  There was a discussion in March of
2001 on perl5-porters where it was suggested that Errno is supposed to
be autoloaded on use of %!, although it appears that mechanism is
still broken.

Note that the build order of extensions needs to be changed so that
Errno is built before modules using AutoLoader.

diff -ur perl-5.6.1.orig/Makefile.SH perl-5.6.1/Makefile.SH
--- perl-5.6.1.orig/Makefile.SH	Fri Apr  6 14:38:46 2001
+++ perl-5.6.1/Makefile.SH	Fri May 11 18:32:55 2001
@@ -300,7 +300,7 @@
 .c$(OBJ_EXT):
 	$(CCCMD) $(PLDLFLAGS) $*.c
 
-all: $(FIRSTMAKEFILE) miniperl extra.pods $(private) $(public) $(dynamic_ext) $(nonxs_ext)
+all: $(FIRSTMAKEFILE) miniperl extra.pods $(private) $(public) $(nonxs_ext) $(dynamic_ext)
 	@echo " ";
 	@echo "	Everything is up to date. 'make test' to run test suite."
 
@@ -805,7 +805,7 @@
 
 # Cannot delegate rebuilding of t/perl to make to allow interlaced
 # test and minitest
-test-prep: miniperl perl preplibrary utilities $(dynamic_ext) $(nonxs_ext) $(TEST_PERL_DLL)
+test-prep: miniperl perl preplibrary utilities $(nonxs_ext) $(dynamic_ext) $(TEST_PERL_DLL)
 	cd t && (rm -f perl$(EXE_EXT); $(LNS) ../perl$(EXE_EXT) perl$(EXE_EXT))
 
 # Second branch is for testing without a tty or controling terminal.
diff -ur perl-5.6.1.orig/ext/DB_File/DB_File.pm perl-5.6.1/ext/DB_File/DB_File.pm
--- perl-5.6.1.orig/ext/DB_File/DB_File.pm	Fri Feb 23 13:57:53 2001
+++ perl-5.6.1/ext/DB_File/DB_File.pm	Fri May 11 18:31:01 2001
@@ -16,6 +16,7 @@
 use warnings;
 use strict;
 use Carp;
+use Errno;
 require Tie::Hash;
 @DB_File::HASHINFO::ISA = qw(Tie::Hash);
 
@@ -212,7 +213,7 @@
     ($constname = $AUTOLOAD) =~ s/.*:://;
     my $val = constant($constname, @_ ? $_[0] : 0);
     if ($! != 0) {
-	if ($! =~ /Invalid/ || $!{EINVAL}) {
+	if ($!{EINVAL}) {
 	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
 	    goto &AutoLoader::AUTOLOAD;
 	}
diff -ur perl-5.6.1.orig/ext/Fcntl/Fcntl.pm perl-5.6.1/ext/Fcntl/Fcntl.pm
--- perl-5.6.1.orig/ext/Fcntl/Fcntl.pm	Fri Feb 23 13:57:54 2001
+++ perl-5.6.1/ext/Fcntl/Fcntl.pm	Fri May 11 18:31:01 2001
@@ -58,6 +58,7 @@
 our($VERSION, @ISA, @EXPORT, @EXPORT_OK, %EXPORT_TAGS, $AUTOLOAD);
 
 require Exporter;
+use Errno;
 use XSLoader ();
 @ISA = qw(Exporter);
 $VERSION = "1.03";
@@ -203,7 +204,7 @@
     (my $constname = $AUTOLOAD) =~ s/.*:://;
     my $val = constant($constname, 0);
     if ($! != 0) {
-	if ($! =~ /Invalid/ || $!{EINVAL}) {
+	if ($!{EINVAL}) {
 	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
 	    goto &AutoLoader::AUTOLOAD;
 	}
diff -ur perl-5.6.1.orig/ext/File/Glob/Glob.pm perl-5.6.1/ext/File/Glob/Glob.pm
--- perl-5.6.1.orig/ext/File/Glob/Glob.pm	Mon Apr  2 15:18:41 2001
+++ perl-5.6.1/ext/File/Glob/Glob.pm	Fri May 11 18:31:01 2001
@@ -6,6 +6,7 @@
     $AUTOLOAD, $DEFAULT_FLAGS);
 
 require Exporter;
+use Errno;
 use XSLoader ();
 require AutoLoader;
 
@@ -86,7 +87,7 @@
     ($constname = $AUTOLOAD) =~ s/.*:://;
     my $val = constant($constname, @_ ? $_[0] : 0);
     if ($! != 0) {
-	if ($! =~ /Invalid/) {
+	if ($!{EINVAL}) {
 	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
 	    goto &AutoLoader::AUTOLOAD;
 	}
diff -ur perl-5.6.1.orig/ext/GDBM_File/GDBM_File.pm perl-5.6.1/ext/GDBM_File/GDBM_File.pm
--- perl-5.6.1.orig/ext/GDBM_File/GDBM_File.pm	Mon Mar 19 19:11:17 2001
+++ perl-5.6.1/ext/GDBM_File/GDBM_File.pm	Fri May 11 18:31:01 2001
@@ -47,6 +47,7 @@
 require Tie::Hash;
 require Exporter;
 use AutoLoader;
+use Errno;
 use XSLoader ();
 @ISA = qw(Tie::Hash Exporter);
 @EXPORT = qw(
@@ -68,7 +69,7 @@
     ($constname = $AUTOLOAD) =~ s/.*:://;
     my $val = constant($constname, @_ ? $_[0] : 0);
     if ($! != 0) {
-	if ($! =~ /Invalid/ || $!{EINVAL}) {
+	if ($!{EINVAL}) {
 	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
 	    goto &AutoLoader::AUTOLOAD;
 	}
diff -ur perl-5.6.1.orig/lib/AutoLoader.pm perl-5.6.1/lib/AutoLoader.pm
--- perl-5.6.1.orig/lib/AutoLoader.pm	Fri Feb 23 13:57:55 2001
+++ perl-5.6.1/lib/AutoLoader.pm	Fri May 11 18:31:01 2001
@@ -1,6 +1,7 @@
 package AutoLoader;
 
 use 5.005_64;
+use Errno;
 our(@EXPORT, @EXPORT_OK, $VERSION);
 
 my $is_dosish;
@@ -247,7 +248,7 @@
         (my $constname = $sub) =~ s/.*:://;
         my $val = constant($constname, @_ ? $_[0] : 0);
         if ($! != 0) {
-            if ($! =~ /Invalid/ || $!{EINVAL}) {
+            if ($!{EINVAL}) {
                 $AutoLoader::AUTOLOAD = $sub;
                 goto &AutoLoader::AUTOLOAD;
             }
diff -ur perl-5.6.1.orig/lib/CPAN.pm perl-5.6.1/lib/CPAN.pm
--- perl-5.6.1.orig/lib/CPAN.pm	Mon Mar 19 19:10:30 2001
+++ perl-5.6.1/lib/CPAN.pm	Fri May 11 18:31:01 2001
@@ -11,6 +11,7 @@
 use Config ();
 use Cwd ();
 use DirHandle;
+use Errno;
 use Exporter ();
 use ExtUtils::MakeMaker (); # $SelfLoader::DEBUG=1;
 use File::Basename ();
@@ -531,7 +532,7 @@
     }
     my $fh;
     unless ($fh = FileHandle->new(">$lockfile")) {
-	if ($! =~ /Permission/) {
+	if ($!{EACCES}) {
 	    my $incc = $INC{'CPAN/Config.pm'};
 	    my $myincc = MM->catfile($ENV{HOME},'.cpan','CPAN','MyConfig.pm');
 	    $CPAN::Frontend->myprint(qq{
diff -ur perl-5.6.1.orig/utils/h2xs.PL perl-5.6.1/utils/h2xs.PL
--- perl-5.6.1.orig/utils/h2xs.PL	Fri Feb 23 13:57:58 2001
+++ perl-5.6.1/utils/h2xs.PL	Fri May 11 18:31:01 2001
@@ -811,8 +811,9 @@
 
 unless( $opt_X || $opt_c || $opt_A ){
 	# we'll have an AUTOLOAD(), and it will have $AUTOLOAD and
-	# will want Carp.
+	# will want Carp and Errno
 	print PM <<'END';
+use Errno;
 use Carp;
 END
 }
@@ -897,7 +898,7 @@
     croak "&$module::constant not defined" if \$constname eq 'constant';
     my \$val = constant(\$constname, \@_ ? \$_[0] : 0);
     if (\$! != 0) {
-	if (\$! =~ /Invalid/ || \$!{EINVAL}) {
+	if (\$!{EINVAL}) {
 	    \$AutoLoader::AUTOLOAD = \$AUTOLOAD;
 	    goto &AutoLoader::AUTOLOAD;
 	}
